doc/wrapper.conf.5
.TH wrapper.conf 5 2013-12-18 "Suid-wrapper 0.9"
.SH NAME
wrapper.conf - Configuration files for build-wrapper

.SH DESCRIPTION
Suid-wrapper provides the possibility to easily build customized setuid wrappers. These wrappers
are built by the build-wrappers tool from configuration files written in CPP. Each wrapper is built
using GCC from information included in its configuration file.
.P
Wrapper configuration files combines two types of directives:
.IP \(bu 4
C macros read by GCC to fill in parameters in the template file
.IP \(bu 4
Comments read by build-wrappers to package the wrapper after its compilation

.SH WRAPPER MACROS
Wrapper variables are defined using the "#define" CPP instruction. They control three kinds of
settings:
.IP \(bu 4
Checks performed by the wrapper on the caller and on program parameters (macros
.BR CHK_* )
.IP \(bu 4
Permissions to which the executed program will be restricted (macros
.BR DO_* )
.IP \(bu 4
Information to execute the program such as path or arguments (macros
.BR EXEC_* )
.P
Valid macros are:
.TP
.B #define CHK_ARGC_MIN <min>
Define the minimum numbers of arguments that may be passed to the wrapper, including the wrapper
name. These arguments will be added to the program arguments after those specified in the
.B EXEC_ARGS
macro. If given, must be included between 1 and 65535.
.TP
.B #define CHK_ARGC_MAX <max>
Define the maximum numbers of arguments that may be passed to the wrapper, including the wrapper
name. These arguments will be added to the program arguments after those specified in the
.B EXEC_ARGS
macro. If given, must be included between 1 and 65535.
.TP
.B #define CHK_ARGS_REGEXP ["<arg1>",[ "<arg2>",[ ... ]]]
Check each wrapper argument against regular expressions. Regexps are separated by commas, each
regexp checks the corresponding argument, excluding the wrapper name. See
.BR regex (7)
for more information about regular expressions.
.TP
.B #define CHK_ARGS_REGEXP_FLAGS <flags>
Apply matching flags to all regular expressions. This permits for example to perform
case-insensitive matchings.
.TP
.B #define CHK_REAL_UID <uid>
Check the real UID of the caller before executing the program. If given, must be included between
1 and 65535.
.TP
.B #define CHK_REAL_GID <gid>
Check the real GID of the caller before executing the program. If given, must be included between
1 and 65535.
.TP
.B #define DO_SETUID <uid>
Set the real UID of the process before calling the program. If given, must be included between
1 and 65535. This option may cause undefined behavior when used with the
.B DO_CAPABILITIES
macro.
.TP
.B #define DO_SETGID <uid>
Set the real GID of the process before calling the program. If given, must be included between
1 and 65535. This option may cause undefined behavior when used with the
.B DO_CAPABILITIES
macro.
.TP
.B #define DO_CAPABILITIES [<cap1> [, <cap2> [, ...]]]
Set capabilities to the given set. Capabilities are separated by commas. This option may cause undefined behavior when used with the
.B DO_SETUID
or
.B DO_SETGID
macros. See
.BR capabilities (7)
for more information about regular expressions.
.TP
.B #define EXEC_PATH "<path>"
Path of the program to execute. This option is
.B mandatory.
.TP
.B #define EXEC_ARGS [arg1 [, arg2 [, ...]]]
Arguments to pass to the program.
.TP
.B #define DEBUG
Enable debugging information. By default, the wrapper silently exits with an EPERM error. With
this macro set, the wrapper will print at which step it failed to perform a specific operation
or check.

.SH BUILD-WRAPPER VARIABLES
These configuration variables are located in the same configuration file as wrapper macros but
must not be interpreted by the preprocessor nor by the compiler. For this reason, they must be
surrounded by C comments marks:
.I "/*"
and
.IR "*/" .
These variables are formatted as follows:
.IP
.B VARIABLE=<value>
.P
Valid variables are:
.TP
.B WRAPPER_UID=(<uid>|<user>)
Set the file UID of the wrapper. By default, the wrapper is owner by root.
.TP
.B WRAPPER_GID=(<gid>|<group>)
Set the file GID of the wrapper. By default, the wrapper has its group set to root.
.TP
.B WRAPPER_PATH=<path>
Set the file location of the wrapper. The wrapper will be moved there after compilation and before
privilege setting. By default, the wrapper is compiled at the path /var/lib/suid-wrapper/$wrapper
.TP
.B WRAPPER_PERMS=<perms>
Set the file permissions of the wrapper. By default, the wrapper has permissions 4750.

.SH NOTES
As the configuration is read by the C preprocessor then the C compiler, the wrapper configuration
file must be a valid C file. As a side effect, C language instructions are allowed in wrapper
macros.
.P
The
.B DO_*
macros does not grant privileges but restrict it. Special privileges are granted by the setuid
bit set on the wrapper file. If you set (or let) the setuid bit on the wrapper with a root owner,
be sure to restrict the privileges afterwards.

.SH FILES
.TP
.B /etc/suid-wrapper/<wrapper>.conf
Wrapper configuration file
.TP
.B /var/lib/suid-wrapper/<wrapper>
Built wrapper

.SH SEE ALSO
.BR build-wrappers (8)

.SH AUTHOR
This manage was written by Emmanuel Thierry <emmanuel.thierry@sekil.fr>, who is the original author of suid-wrapper.


wrapper/wrapper.c
/**
 * Copyright (c) 2013, Emmanuel Thierry <emmanuel.thierry@sekil.fr>
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice,
 * this list of conditions and the following disclaimer.
 *
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 * this list of conditions and the following disclaimer in the documentation
 * and/or other materials provided with the distribution.
 *
 * 3. Neither the name of the Emmanuel Thierry nor the names of its
 * contributors may be used to endorse or promote products derived from this
 * software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 *
 */


#include <errno.h>
#include <stdlib.h>

#ifdef DEBUG
#include <stdio.h>
#endif

#ifdef CHK_ARGS_REGEXP
#include <string.h>
#include <regex.h>
#include <sys/types.h>
#endif

#ifdef DO_CAPABILITIES
#include <sys/capability.h>
#include <sys/prctl.h>
#include <sys/types.h>
#include <linux/securebits.h>
#endif



#ifdef DEBUG
#define EXIT_CHK(error) \
	{fprintf(stderr, "Failed to pass %s test with errno %d\n", error, errno); exit(EPERM);}
#define EXIT_DO(error) \
	{fprintf(stderr, "Failed to grant %s privilege with errno %d\n", error, errno); exit(EPERM);}
#else
#define EXIT_CHK(error) \
	exit(EPERM)
#define EXIT_DO(error) \
	exit(EPERM)
#endif

#ifdef CHK_ARGS_REGEXP
const char const *args_regexp[] = { CHK_ARGS_REGEXP };
#define _ARG_MAX (sizeof(args_regexp) / sizeof(char *))
#endif

#ifdef DO_CAPABILITIES
cap_value_t capabilities[] = { DO_CAPABILITIES };
#define _CAP_MAX (sizeof(capabilities) / sizeof(cap_value_t))
#endif

char * const exec_args[] = { EXEC_ARGS };
#define _EXEC_MAX (sizeof(exec_args) / sizeof(char *))


int main(int argc, char * const argv[], char * const envp[])
{
	char *new_argv[_EXEC_MAX + argc + 1];
	int i;
#ifdef DO_CAPABILITIES
	cap_t caps, empty;
#endif


#ifdef CHK_WRAPPER_PATH
	if (strcmp(argv[0], CHK_WRAPPER_PATH))
		EXIT_CHK("CHK_WRAPPER_PATH");
#endif

#ifdef CHK_ARGC_MIN
	if (argc < CHK_ARGC_MIN)
		EXIT_CHK("CHK_ARGC_MIN");
#endif

#ifdef CHK_ARGC_MAX
	if (argc > CHK_ARGC_MAX)
		EXIT_CHK("CHK_ARGC_MAX");
#endif

#ifdef CHK_REAL_UID
	if (getuid() != CHK_REAL_UID)
		EXIT_CHK("CHK_REAL_UID");
#endif

#ifdef CHK_REAL_GID
	if (getgid() != CHK_REAL_GID)
		EXIT_CHK("CHK_REAL_GID");
#endif

	new_argv[0] = EXEC_PATH;
	for (i=1; i<_EXEC_MAX+1; i++) {
		new_argv[i] = exec_args[i-1];
	}
	for (; i<_EXEC_MAX+argc; i++) {
#ifdef CHK_ARGS_REGEXP
		if ((i < _EXEC_MAX + _ARG_MAX + 1) && args_regexp[i-_EXEC_MAX-1]) {
			regex_t preg;
			memset(&preg, 0, sizeof(preg));
			if (regcomp(&preg, args_regexp[i-_EXEC_MAX-1], CHK_ARGS_REGEXP_FLAGS))
				EXIT_CHK("CHK_ARGS_REGEXP(regcomp)");
			if (regexec(&preg, argv[i-_EXEC_MAX], 0, NULL, 0))
				EXIT_CHK("CHK_ARGS_REGEXP(regexec)");
			regfree(&preg);
		}
#endif
		new_argv[i] = argv[i-_EXEC_MAX];
	}
	new_argv[i] = NULL;

#ifdef DO_CAPABILITIES
	if (setuid(0))
		EXIT_DO("DO_CAPABILITIES(setuid)");
	if (prctl(PR_SET_SECUREBITS, SECBIT_KEEP_CAPS, 0, 0, 0))
		EXIT_DO("DO_CAPABILITIES(prctl)");
	caps = cap_init();
	if (cap_set_flag(caps, CAP_EFFECTIVE, _CAP_MAX, capabilities, CAP_SET) ||
	    cap_set_flag(caps, CAP_PERMITTED, _CAP_MAX, capabilities, CAP_SET) ||
	    cap_set_flag(caps, CAP_INHERITABLE, _CAP_MAX, capabilities, CAP_SET) ||
	    cap_set_proc(caps))
		EXIT_DO("DO_CAPABILITIES(cap_set_proc");
	cap_free(caps);
#endif

#ifdef DO_SETGID
	if (setgid(DO_SETGID))
		EXIT_DO("DO_SETGID");
#endif

#ifdef DO_SETUID
	if (setuid(DO_SETUID))
		EXIT_DO("DO_SETUID");
#endif

#ifdef DEBUG
	fprintf(stderr, "All checks passed\n");
	fprintf(stderr, "%s", EXEC_PATH);
	for (i=0; new_argv[i]; i++)
		fprintf(stderr, " \"%s\"", new_argv[i]);
	fprintf(stderr, "\n");
#endif
	execv(EXEC_PATH, new_argv);
}
